Change capture_scalar_outputs to use SymInt/SymFloat rather than Tensor to model scalars

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93150

Previously, Dynamo faked support for item() when `capture_scalar_outputs` was True by representing it internally as a Tensor. With dynamic shapes, this is no longer necessary; we can represent it directly as a SymInt/SymFloat. Do so. Doing this requires you to use dynamic shapes; in principle we could support scalar outputs WITHOUT dynamic shapes but I won't do this unless someone hollers for it.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

Differential Revision: [D42885775](https://our.internmc.facebook.com/intern/diff/D42885775)

ezyang committed:
When dynamic shapes enabled, use default handling for item

The fake implementation for item produces an unbacked SymInt/SymFloat
so there is no need for Dynamo to special case it.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "When dynamic shapes enabled, use default handling for item"


The fake implementation for item produces an unbacked SymInt/SymFloat
so there is no need for Dynamo to special case it.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "When dynamic shapes enabled, use default handling for item"


The fake implementation for item produces an unbacked SymInt/SymFloat
so there is no need for Dynamo to special case it.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "When dynamic shapes enabled, use default handling for item"


The fake implementation for item produces an unbacked SymInt/SymFloat
so there is no need for Dynamo to special case it.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
OK, so, uh, I can't actually land this, because although I can get the graph all the way to inductor, inductor doesn't know how to compile graphs with item calls. I guess I'll gate this behind another config as well.

ezyang committed:
Update on "When dynamic shapes enabled, use default handling for item"


The fake implementation for item produces an unbacked SymInt/SymFloat
so there is no need for Dynamo to special case it.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Change capture_scalar_outputs to use SymInt/SymFloat rather than Tensor to model scalars"


Previously, Dynamo faked support for item() when `capture_scalar_outputs` was True by representing it internally as a Tensor. With dynamic shapes, this is no longer necessary; we can represent it directly as a SymInt/SymFloat. Do so. Doing this requires you to use dynamic shapes; in principle we could support scalar outputs WITHOUT dynamic shapes but I won't do this unless someone hollers for it.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Change capture_scalar_outputs to use SymInt/SymFloat rather than Tensor to model scalars"


Previously, Dynamo faked support for item() when `capture_scalar_outputs` was True by representing it internally as a Tensor. With dynamic shapes, this is no longer necessary; we can represent it directly as a SymInt/SymFloat. Do so. Doing this requires you to use dynamic shapes; in principle we could support scalar outputs WITHOUT dynamic shapes but I won't do this unless someone hollers for it.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

albanD reviewed (commented):
Sounds good to me but I'll let Voz take a look for the dynamo code

ezyang committed:
Update on "Change capture_scalar_outputs to use SymInt/SymFloat rather than Tensor to model scalars"


Previously, Dynamo faked support for item() when `capture_scalar_outputs` was True by representing it internally as a Tensor. With dynamic shapes, this is no longer necessary; we can represent it directly as a SymInt/SymFloat. Do so. Doing this requires you to use dynamic shapes; in principle we could support scalar outputs WITHOUT dynamic shapes but I won't do this unless someone hollers for it.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Change capture_scalar_outputs to use SymInt/SymFloat rather than Tensor to model scalars"


Previously, Dynamo faked support for item() when `capture_scalar_outputs` was True by representing it internally as a Tensor. With dynamic shapes, this is no longer necessary; we can represent it directly as a SymInt/SymFloat. Do so. Doing this requires you to use dynamic shapes; in principle we could support scalar outputs WITHOUT dynamic shapes but I won't do this unless someone hollers for it.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42885775).

eellison commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -319,22 +319,16 @@ def call_method(
+            and config.capture_scalar_outputs
+            and not config.dynamic_shapes
+        ):
+            raise AssertionError(
```
Is it necessary to throw here ?

voznesenskym replied:
Yes - misconfiguration should throw.

voznesenskym commented on test/dynamo/test_export.py:
```diff
@@ -320,6 +320,7 @@ def func(x, z, k):
 
         self.assertTrue(torch._dynamo.utils.same(real_result, dynamo_result))
 
+    @patch.object(torch._dynamo.config, "dynamic_shapes", True)
```
this is already run under test/dynamo/test_dynamic_shapes.py, unless you want this to only run under dynamic shapes?

voznesenskym replied:
Saw rest of pr, sg

ezyang replied:
There is a problem which is when we mark things as dynamic shapes True here, we actually are technically running the test twice but under same config. This is annoying to fix though, and is preexisting condition

ezyang commented:
@pytorchbot merge -f "flaky rocm only"
